src/kubernetes.cc

```diff
@@ -907,6 +980,93 @@ json::value KubernetesReader::FindTopLevelController(
+
+  std::lock_guard<std::recursive_mutex> lock(service_mutex_);
+  auto service_it = service_to_metadata_.find(encoded_ref);
+  if (is_deleted) {
+    if (service_it != service_to_metadata_.end()) {
+      service_to_metadata_.erase(service_it);
+    }
+  } else {
+    json::value service_value = service->Clone();
+    if (service_it == service_to_metadata_.end()) {
+      service_to_metadata_.emplace(encoded_ref, std::move(service_value));
+    } else {
+      service_it->second = std::move(service_value);
+    }
+  }
+}
+
+void KubernetesReader::UpdateServiceToPodsCache(
+    const json::Object* endpoints, bool is_deleted) throw(json::Exception) {
+#ifdef VERBOSE
+  LOG(DEBUG) << "UpdateServiceToPodsCache(" << *endpoints << ")";
+#endif
+  const json::Object* metadata = endpoints->Get<json::Object>("metadata");
+  const std::string namespace_name = metadata->Get<json::String>("namespace");
+  // Endpoints name is same as the matching service name.
+  const std::string service_name = metadata->Get<json::String>("name");
+  const std::string encoded_ref = boost::algorithm::join(
+      std::vector<std::string>{namespace_name, service_name}, "/");
+  std::vector<std::string> pod_names;
+
+  // Only extract the pod names when this is not a deletion. In the case of
+  // a deletion, we delete the mapping below.
+  if (!is_deleted && endpoints->Has("subsets") &&
+      !endpoints->at("subsets")->Is<json::Null>()) {
+    const json::Array* subsets = endpoints->Get<json::Array>("subsets");
+    for (const json::value& subset : *subsets) {
+      const json::Object* subset_obj = subset->As<json::Object>();
+      if (!subset_obj->Has("addresses")) {
+        continue;
+      }
+      const json::Array* addresses = subset_obj->Get<json::Array>("addresses");
+      for (const json::value& address : *addresses) {
+        const json::Object* address_obj = address->As<json::Object>();
+        if (!address_obj->Has("targetRef")) {
+          continue;
+        }
+        const json::Object* ref = address_obj->Get<json::Object>("targetRef");
+        if (!(ref->Has("kind") && ref->Has("name"))) {
+          continue;
+        }
+        const std::string& target_kind = ref->Get<json::String>("kind");
+        if (target_kind != "Pod") {
+          LOG(INFO) << "Found a resource other than a pod in Endpoint "
+                    << service_name << "'s targetRef: " << target_kind;
+          continue;
+        }
+        const std::string& pod_name = ref->Get<json::String>("name");
+        pod_names.push_back(pod_name);
+      }
+    }
+  }
+
+  std::lock_guard<std::recursive_mutex> lock(service_mutex_);
+  auto service_it = service_to_pods_.find(encoded_ref);
+  if (is_deleted) {
+    if (service_it != service_to_pods_.end()) {
+      service_to_pods_.erase(service_it);
+    }
+  } else {
+    if (service_it == service_to_pods_.end()) {
+      service_to_pods_.emplace(encoded_ref, pod_names);
+    } else {
+      service_it->second = pod_names;
```

Comment:
[Optional] If you're willing to hold the lock a little longer, you could simplify the code a bit and avoid these copies by adding the pod names directly to the vector in the map:
```cpp
  const std::string encoded_ref = boost::algorithm::join(
      std::vector<std::string>{namespace_name, service_name}, "/");
  std::lock_guard<std::recursive_mutex> lock(service_mutex_);
  if (is_deleted) {
    service_to_pods_.erase(encoded_ref);
    return;
  }
  auto it_inserted = service_to_pods_.emplace(encoded_ref, {});
  std::vector<std::string>& pod_names = it_inserted.first->second;
  pod_names.clear();
  if (endpoints->Has("subsets") &&
      !endpoints->at("subsets")->Is<json::Null>()) {
    ...
    pod_names.push_back(pod_name);
    ...
  }
```
This will make insertion faster, at the cost of holding up queries (when they happen) until the update is completed. Since we expect to be modifying this data a lot more often than querying it, this tradeoff makes sense to me, but I'll leave it up to you.

Reply:
Ah, you're right. I forgot that we have to re-construct the k8s_cluster resource each time. Oh, well.
I _am_ wondering whether the cost of copying large vectors around would be comparable with the cost of constructing the vector in-place, but we can defer it to later (or, optionally, add a TODO to investigate).